osprepare/main.go

```diff
@@ -0,0 +1,118 @@
+//
+// Copyright © 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package osprepare
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// Minimal versions supported by ciao
+const (
+	MinDockerVersion = "1.11.0"
+	MinQemuVersion   = "2.5.0"
+)
+
+// PackageRequirement contains the BinaryName expected to
+// exist on the filesystem once PackageName is installed
+// (e.g: { '/usr/bin/qemu-system-x86_64', 'qemu'})
+type PackageRequirement struct {
+	BinaryName  string
+	PackageName string
+}
+
+// PackageRequirements type allows to create complex
+// mapping to group a set of PackageRequirement to a single
+// key.
+// (e.g:
+//
+//	"ubuntu": {
+//		{"/usr/bin/docker", "docker"},
+//	},
+//	"clearlinux": {
+//		{"/usr/bin/docker", "containers-basic"},
+//	},
+// )
+type PackageRequirements map[string][]*PackageRequirement
+
+// BootstrapRequirements lists required dependencies for absolutely core
+// functionality across all Ciao components
+var BootstrapRequirements = PackageRequirements{
+	"ubuntu": {
+		{"/usr/bin/cephfs", "ceph-fs-common"},
+		{"/usr/bin/ceph", "ceph-common"},
+	},
+	"clearlinux": {
+		{"/usr/bin/ceph", "storage-cluster"},
+	},
+}
+
+// CollectPackages returns a list of non-installed packages from
+// the PackageRequirements received
+func collectPackages(dist distro, reqs PackageRequirements) []string {
+	// For now just support keys like "ubuntu" vs "ubuntu:16.04"
+	var pkgsMissing []string
+	if reqs == nil {
+		return nil
+	}
+
+	id := dist.getID()
+	if pkgs, success := reqs[id]; success {
+		for _, pkg := range pkgs {
+			// Have the path existing, skip.
+			if pathExists(pkg.BinaryName) {
+				continue
+			}
+			// Mark the package for installation
+			pkgsMissing = append(pkgsMissing, pkg.PackageName)
+		}
+		return pkgsMissing
+	}
+	return nil
+}
+
+// InstallDeps installs all the dependencies defined in a component
+// specific PackageRequirements in order to enable running the component
+func InstallDeps(reqs PackageRequirements) {
```

Comment:
We need to add a context parameter here so we can cancel the InstallDeps function.  Otherwise launcher will not be able to clean up correctly when asked to exit by the OS.  I'm guessing InstallDeps is a long running function.